packages/flame/lib/src/components/component.dart

```diff
@@ -435,38 +550,30 @@ class Component {
     return Future.wait(futures);
   }
 
-  /// Adds this component to the provided [parent] (see [add] for details).
+  /// Adds this component as a child of [parent] (see [add] for details).
   Future<void>? addToParent(Component parent) {
     assert(
       _parent == null,
       '$this cannot be added to $parent because it already has a parent: '
       '$_parent',
     );
-    assert(
-      _state == LifecycleState.uninitialized ||
-          _state == LifecycleState.removed,
-    );
     _parent = parent;
     parent.lifecycle._children.add(this);
-
-    if (!isLoaded) {
-      final root = parent.findGame();
-      if (root != null) {
-        assert(
-          root.hasLayout,
-          'add() called before the game has a layout. Did you try to add '
-          'components from the constructor? Use the onLoad() method instead.',
-        );
-        return _load();
-      }
+    if (!isLoaded && (parent.findGame()?.hasLayout ?? false)) {
+      return _startLoading();
     }
     return null;
   }
 
-  /// Removes a component from the component tree, calling [onRemove] for it and
-  /// its children. The [onRemove] handler will only be called if there was an
-  /// [onMount] call previously, i.e. when removing components that are properly
-  /// mounted.
+  /// Removes a component from the component tree.
+  ///
+  /// This will call [onRemove] for the component and its children, but only if
+  /// there was an [onMount] call previously, i.e. when removing a component
+  /// that was properly mounted.
+  ///
+  /// A component can be removed even before it finishes mounting, however such
+  /// component cannot be added back into the tree until it at least finishes
+  /// loading.
```

Comment:
Couldn't we place it straight into the loading list if it is only halfway through loading when it is re-added?

Reply:
I'm worried about the case when the user is accessing `parent` from `onLoad`, and then suddenly that parent would change or become null in the middle of their code.
Of course, accessing `parent` from `onLoad` is a kind of an anti-pattern, but we have been allowing it so far, so not sure how to handle this.

We could, for example, make it an error to access `parent` while the component `isLoading` -- but that would be a breaking change, and has to be done in a separate PR, after careful consideration.

Reply:
Right, I think that could be a good change so that people move such logic to `onMount` instead.
But as you said, outside of the scope of this PR then.